src/Lib.P2P/Multiplex/Substream.cs

```diff
@@ -0,0 +1,217 @@
+        /// </value>
+        public PacketType SentMessageType = PacketType.MessageReceiver;
+
+        /// <summary>
+        ///   The stream identifier.
+        /// </summary>
+        /// <value>
+        ///   The session initiator allocates odd IDs and the session receiver allocates even IDs.
+        /// </value>
+        public long Id;
+
+        /// <summary>
+        ///   A name for the stream.
+        /// </summary>
+        /// <value>
+        ///   Names do not need to be unique.
+        /// </value>
+        public string Name;
+
+        /// <summary>
+        ///   The multiplexor associated with the substream.
+        /// </summary>
+        public Muxer Muxer { get; set; }
+
+        /// <inheritdoc />
+        public override bool CanRead => !eos;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => outStream != null;
+
+        /// <inheritdoc />
+        public override bool CanTimeout => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+
+        /// <summary>
+        ///   Add some data that should be read by the stream.
+        /// </summary>
+        /// <param name="data">
+        ///   The data to be read.
+        /// </param>
+        /// <remarks>
+        ///   <b>AddData</b> is called when the muxer receives a packet for this
+        ///   stream.
+        /// </remarks>
+        public void AddData(byte[] data) { inBlocks.Post(data); }
+
+        /// <summary>
+        ///   Indicates that the stream will not receive any more data.
+        /// </summary>
+        /// <seealso cref="AddData(byte[])"/>
+        /// <remarks>
+        ///   <b>NoMoreData</b> is called when the muxer receives a packet to
+        ///   close this stream.
+        /// </remarks>
+        public void NoMoreData() { inBlocks.Complete(); }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+#pragma warning disable VSTHRD002 
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task<int> ReadAsync(byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            var total = 0;
+            while (count > 0 && !eos)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'while' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)